### PR TITLE
feat/show history of documents sent to parties (FPLA-1173)

### PR DESCRIPTION
**Pull request by tomaszpowroznik**
https://tools.hmcts.net/jira/browse/FPLA-1173

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X ] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DocumentSenderService.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    private final Clock clock;
+    private final DateFormatterService dateFormatterService;
+
+    public List<SentDocument> send(DocumentReference mainDocument,
+                                   Long ccdCaseNumber,
+                                   String familyManCaseNumber,
+                                   List<Representative> representativesServedByPost) {
+        //TODO  generate, stitch and send here
+
+        List<SentDocument> printedDocuments = new ArrayList<>();
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DocumentSenderService.java]: let's change variable to match type: sentDocuments

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/SendDocumentController.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+import java.util.List;
+
+import static uk.gov.hmcts.reform.fpl.enums.RepresentativeServingPreferences.POST;
+
+@Api
+@RestController
+@RequestMapping("/callback/send-document")
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class SendDocumentController {
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/SendDocumentController.java]: when is the best time to write integration test for this? as part of this PR or when we have everything together? need to check that documentsSentToParties is updated in CaseData

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DocumentSenderService.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import static java.time.LocalDateTime.now;
+
+@Service
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class DocumentSenderService {
+
+    private final Clock clock;
+    private final DateFormatterService dateFormatterService;
+
+    public List<SentDocument> send(DocumentReference mainDocument,
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DocumentSenderService.java]: Something that needs to be tested but again probably needs to wait until we have stitching service etc. We could probably do a draft for test perhaps

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/SentDocument.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import lombok.Builder;
+import lombok.Data;
+import uk.gov.hmcts.reform.fpl.model.common.DocumentReference;
+
+@Data
+@Builder(toBuilder = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class SentDocument {
```
[the remark this answers is not in the document]
Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/SentDocument.java]: Done, changed to DocumentSentToParty.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/SentDocumentHistoryService.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+        List<Element<DocumentsSentToParty>> historicalRecords = defaultIfNull(sentDocumentsHistory, new ArrayList<>());
+
+        printedDocuments.forEach(printedDocument -> addToHistory(historicalRecords, printedDocument));
+
+        return historicalRecords;
+    }
+
+    private void addToHistory(List<Element<DocumentsSentToParty>> documentsSentToPartyCollection,
+                              SentDocument printedDocument) {
```
[the remark this answers is not in the document]
Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/SentDocumentHistoryService.java]: done

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/SendDocumentEventControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Test
+    void shouldSendDocumentToRepresentativesWithPostServingPreferences() {
+
+        Representative representative1 = representative("John Smith", POST);
+        Representative representative2 = representative("Alex Brown", EMAIL);
+        Representative representative3 = representative("Emma White", DIGITAL_SERVICE);
+
+        DocumentReference documentToBeSend = testDocument();
```
[the remark this answers is not in the document]
Comment (tomaszpowroznik) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/SendDocumentEventControllerTest.java]: done

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DocumentSenderService.java (hunk 1 above)
Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DocumentSenderService.java]: done

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/SendDocumentController.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    private final DocumentSenderService documentSenderService;
+    private final SentDocumentHistoryService documentHistoryService;
+    private final RepresentativeService representativeService;
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSave(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        List<Representative> representativesServedByPost =
```
[the remark this answers is not in the document]
Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/SendDocumentController.java]: done

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/SendDocumentController.java (hunk 2 above)
Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/SendDocumentController.java]: there is test for it, you even commented on it, see SendDocumentEventControllerTest

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DocumentSenderService.java (hunk 3 above)
Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DocumentSenderService.java]: i guess, Piotr story should do this, anyway added test coverage for what is there already

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/DocumentSenderServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    private static DocumentSentToParty documentSentToRepresentative(DocumentReference documentReference,
+                                                                    Representative representative) {
+        return DocumentSentToParty.builder()
+            .partyName(representative.getFullName())
+            .document(documentReference)
+            .sentAt(FORMATTED_SENT_TIME)
+            .build();
+    }
+
+    private static Representative testRepresentative() {
```
Comment (Gogetter) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/DocumentSenderServiceTest.java]: we could make this and `documentSentToRepresentative` to be non-static

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/SentDocumentHistoryServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.unwrapElements;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.wrapElements;
+import static uk.gov.hmcts.reform.fpl.utils.TestDataHelper.testDocument;
+
+public class SentDocumentHistoryServiceTest {
+
+    private static final String PARTY_1 = "John Smith";
+    private static final String PARTY_2 = "Alex Green";
+
+    SentDocumentHistoryService sentDocumentHistoryService = new SentDocumentHistoryService();
```
Comment (Gogetter) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/SentDocumentHistoryServiceTest.java]: would another test class be extending `SentDocumentHistoryServiceTest` which would need `sentDocumentHistoryService`?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/SentDocumentHistoryServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    private static DocumentSentToParty sentDocument(String partyName) {
+        return DocumentSentToParty.builder()
+            .partyName(partyName)
+            .document(testDocument())
+            .build();
+    }
+
+    private static DocumentsSentToParty sentDocumentsHistory(String partyName, DocumentSentToParty... sentDocuments) {
```
Comment (Gogetter) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/SentDocumentHistoryServiceTest.java]: we could make this and `sentDocument` to be non-static

File: service/src/test/java/uk/gov/hmcts/reform/fpl/utils/TestDataHelper.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
+package uk.gov.hmcts.reform.fpl.utils;
+
+import uk.gov.hmcts.reform.fpl.model.common.DocumentReference;
+
+import static org.apache.commons.lang3.RandomStringUtils.randomAlphanumeric;
+
+public class TestDataHelper {
```
Comment (Gogetter) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/utils/TestDataHelper.java]: 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/SendDocumentController.java (hunk 2 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/SendDocumentController.java]: oh I am blind 🤦‍♂

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/DocumentSenderServiceTest.java (hunk 8 above)
Comment (tomaszpowroznik) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/DocumentSenderServiceTest.java]: not sure what we gain making it non static. I did it static on purpose, to indicate it does not rely on current object state

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/SentDocumentHistoryServiceTest.java (hunk 9 above)
Comment (tomaszpowroznik) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/SentDocumentHistoryServiceTest.java]: rather not, i dont think that extending test class is good idea

File: ccd-definition/CaseEvent/AnyState.json [hunk 12]
```diff
@@ -0,0 +1,12 @@
+[
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "ID": "internal-change:SEND-DOCUMENT",
```
Comment (swalker125) [inline ccd-definition/CaseEvent/AnyState.json]: hmm I am not sure internal-change is right here. I believe we wanted to use this phrasing for when there is an internal change of state. I would suggest something like internal-event? Thoughts?

**Pull request closed without merging**